### PR TITLE
Added consumers and reconnect.

### LINKED ISSUE #1: Add consumers.

### PR DESCRIPTION
**Pull request by cheshir**
Close #1

### CODE REVIEW

File: consumer.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+			return
+		}
+	}
+}
+
+// Force stop.
+// TODO Add wait group.
+func (worker worker) Stop() {
+	isNeedToShutdown := atomic.CompareAndSwapInt32(&worker.status, statusRunning, statusStopped)
+	if isNeedToShutdown {
```
Comment (ynotLeft) [inline consumer.go]: Consider renaming `isNeedToShutdown` to `needsToShutdown`.
This way variable name will answer a Yes/No question and be a valid English construct if used.

File: mq.go [hunk 2]
```diff
@@ -104,22 +131,64 @@ func (mq *MQ) setupProducers(producers Producers) error {
...
+	if _, ok := mq.consumers.Get(config.Name); ok {
+		return fmt.Errorf(`Consumer with name "%s" is already registered`, config.Name)
+	}
+
+	// Consumer must have at least one worker.
+	if config.Workers == 0 {
+		config.Workers = 1
+	}
+
+	consumer := newConsumer(config) // We need to save whole config for reconnection.
```
Comment (ynotLeft) [inline mq.go]: We need to save a whole config for reconnect.

File: mq.go [hunk 3]
```diff
@@ -104,22 +131,64 @@ func (mq *MQ) setupProducers(producers Producers) error {
...
+
+	for i := 0; i < config.Workers; i++ {
+		deliveries, err := mq.channel.Consume(consumer.queue, "", consumer.options)
+		if err != nil {
+			return fmt.Errorf(`Error during starting consumer %s: %s`, config.Name, err)
```
Comment (ynotLeft) [inline mq.go]: Error during the consumer starting.
or 
Error during the starting of consumer.

File: mq.go [hunk 4]
```diff
@@ -104,22 +131,64 @@ func (mq *MQ) setupProducers(producers Producers) error {
...
+		}
+
+		consumer.workers[i] = newWorker(deliveries)
+	}
+
+	mq.consumers.Set(config.Name, consumer) // Workers will start after calling method consumer.Consume.
```
Comment (ynotLeft) [inline mq.go]: Workers will start after consumer.Consume method call.

File: mq.go [hunk 5]
```diff
@@ -137,48 +206,145 @@ func (mq *MQ) connect() error {
...
 	mq.connection = connection
 	mq.channel = channel
 
+	go mq.handleCloseEvent()
+
 	return nil
 }
 
+// Register close handler.
+// To more details visit https://godoc.org/github.com/streadway/amqp#Connection.NotifyClose.
```
Comment (ynotLeft) [inline mq.go]: For more details 
or 
To get more details

File: mq.go [hunk 6]
```diff
@@ -137,48 +206,145 @@ func (mq *MQ) connect() error {
...
+}
+
 func (mq *MQ) errorHandler() {
 	for err := range mq.internalErrorChannel {
 		select {
-		case mq.errorChannel <- err:
-			mq.processError(err)
+		case mq.errorChannel <- err: // Proxies errors to the user.
 		default: // Drop errors if channel buffer is full.
+			// TODO probably it makes sense to make it optional or even remove.
```
Comment (ynotLeft) [inline mq.go]: It probably makes sense to make it optional or even remove.

File: mq.go [hunk 7]
```diff
@@ -137,48 +206,145 @@ func (mq *MQ) connect() error {
...
 }
 
+// Reconnect stops current producers and consumers,
+// recreates connection to the rabbit and than runs producers and consumers.
 func (mq *MQ) reconnect() {
-	mq.connect()
-	mq.producers.SetNewChannel(mq.channel)
+	notBusy := atomic.CompareAndSwapInt32(&mq.reconnectStatus, statusReadyForReconnect, statusReconnecting)
+	if !notBusy {
+		// There is no need to start new reconnect if previous is not finished yet.
```
Comment (ynotLeft) [inline mq.go]: There is no need to start a new reconnect if the previous one is not finished yet.

File: producer.go [hunk 8]
```diff
@@ -28,15 +46,37 @@ func (producer *producer) setChannel(channel wabbit.Channel) {
...
+
+	for {
+		select {
+		case message := <-producer.publishChannel:
+			err := producer.produce(message)
+			if err != nil {
+				producer.errorChannel <- err
+
+				// TODO make it optional.
+				// TODO check error types. Why message is broken?
```
Comment (ynotLeft) [inline producer.go]: Why the message is broken?

File: producer.go [hunk 9]
```diff
@@ -45,3 +85,12 @@ func (producer *producer) produce(message []byte) error {
 
 	return producer.channel.Publish(producer.exchange, producer.routingKey, message, producer.options)
 }
+
+// Stops the worker if it is running.
+// TODO Add wait group.
+func (producer *producer) Stop() {
+	isNeedToShutdown := atomic.CompareAndSwapInt32(&producer.status, statusRunning, statusStopped)
```
Comment (ynotLeft) [inline producer.go]: consider renaming to needsToShutdown

**Comment by cheshir** [comment]
@ynotLeft thank you for your help.

**Pull request merged**